File: easy/ascii_art.py
import sys
import math
# To debug: print("Debug messages...", file=sys.stderr, flush=True)

l = int(input())
h = int(input())
t = input()

rows = []
alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
rows_ascii = []

for i in range(h):
    row = input()
    rows.append(row)

to_ascii = [c for c in t]

for lign in range(h):
    row_ascii = ''
    for j in range(len(to_ascii)):
        char = to_ascii[j].upper()
        if alphabet.find(char) != -1:
            letter_index = alphabet.find(char)
            row_ascii += rows[lign][slice((letter_index * l),(letter_index * l + l), 1)]
        else:
            interrogation_index = len(rows[lign])
            row_ascii += rows[lign][slice((interrogation_index) - l,(interrogation_index), 1)]

    rows_ascii.append(row_ascii)

for row_ascii in rows_ascii:
    row_ascii += "\n"

response = "\n".join(rows_ascii)

print(response)

File: easy/ascii_art.rb
l = gets.to_i
h = gets.to_i
t = gets.chomp

rows = []
response = ""

alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

h.times do
  row = gets.chomp
  rows.push(row)
end

to_ascii = t.split('')

rows_ascii = []

for i in 0...h
    row_ascii = ""
    to_ascii.each do |char|
        # test_char = char.match?('[a-zA-Z]').to_s
        char = char.upcase
        test_char = alphabet.include?(char)
        if test_char
            char = char.upcase
            letter_index = alphabet.index(char)
            row_ascii += rows[i].slice((letter_index * l)...(letter_index * l + l))
        else
            # STDERR.puts "in the else"
            interrogation_index = rows[i].length
            row_ascii += rows[i].slice((interrogation_index - l)...(interrogation_index))
        end
    end
    rows_ascii.push(row_ascii)
end
response = rows_ascii.join("\n")
puts response
